sopel/irc/backends.py

```diff
@@ -0,0 +1,252 @@
+        self.timeout_scheduler.add_job(timeout_job)
+
+    def run_forever(self):
+        asyncore.loop()
+
+    def initiate_connect(self, host, port, source_address):
+        self.host = host
+        self.port = port
+        self.source_address = source_address
+
+        LOGGER.info('Connecting to %s:%s...', host, port)
+        try:
+            LOGGER.debug('Set socket')
+            self.set_socket(socket.create_connection((host, port),
+                            source_address=source_address))
+            LOGGER.debug('Connection attempt')
+            self.connect((host, port))
+        except socket.error as e:
+            LOGGER.exception('Connection error: %s', e)
+            self.handle_close()
+
+    def handle_connect(self):
+        """Called when the active opener's socket actually makes a connection."""
+        LOGGER.info('Connection accepted by the server...')
+        self.timeout_scheduler.start()
+        self.bot.on_connect()
+
+    def handle_close(self):
+        """Called when the socket is closed"""
+        self.timeout_scheduler.stop()
+        self.timeout_scheduler.join(timeout=15)
+
+        LOGGER.info('Connection closed...')
+        try:
+            self.bot.on_close()
+        finally:
+            if self.socket:
+                LOGGER.debug('Closing socket')
+                self.close()
+                LOGGER.info('Closed!')
+
+    def handle_error(self):
+        """Called when an exception is raised and not otherwise handled."""
+        LOGGER.info('Connection error...')
+        self.bot.on_error()
+
+    def collect_incoming_data(self, data):
+        # We can't trust clients to pass valid unicode.
+        try:
+            data = unicode(data, encoding='utf-8')
+        except UnicodeDecodeError:
+            # not unicode, let's try cp1252
+            try:
+                data = unicode(data, encoding='cp1252')
+            except UnicodeDecodeError:
+                # Okay, let's try ISO8859-1
+                try:
+                    data = unicode(data, encoding='iso8859-1')
+                except UnicodeDecodeError:
+                    # Discard line if encoding is unknown
+                    return
+        if data:
+            self.bot.log_raw(data, '<<')
+        self.buffer += data
+        self.last_event_at = datetime.datetime.utcnow()
+
+    def found_terminator(self):
+        line = self.buffer
+        if line.endswith('\r'):
+            line = line[:-1]
+        self.buffer = ''
+        self.bot.on_message(line)
+
+    def on_scheduler_error(self, scheduler, exc):
+        """Called when the Job Scheduler fails."""
+        LOGGER.exception('Error with the timeout scheduler: %s', exc)
+        self.handle_close()
+
+    def on_job_error(self, scheduler, job, exc):
+        """Called when a job from the Job Scheduler fails."""
+        LOGGER.exception('Error with the timeout scheduler: %s', exc)
+        self.handle_close()
```

Comment:
Hm. Indeed that's weird. Why did I do that?

Reply:
Oh, I know. It's because on the backend, there is one and only one job scheduler: the timeout one. So when there is a timeout error, the bot force-closes the connection.

This is the current behavior in 6.x & master.